evap/student/templates/student_vote.html

```diff
@@ -308,6 +308,88 @@ <h3 class="mb-3">{{ evaluation.full_name }} ({{ evaluation.course.semester.name
                 $("#text_results_publish_confirmation_top").prop('checked', $("#text_results_publish_confirmation_bottom").prop('checked'));
             }
+
+            const things_to_decide = [];
+            for (const form of document.getElementsByClassName("vote-form")) {
+                for (const row of form.getElementsByClassName("row")) {
+                    const selectables = row.getElementsByClassName("vote-selectable");
+                    things_to_decide.push(Array.from(selectables));
+                }
+            }
+
+            things_to_decide.push([document.getElementById("presave-btn"), document.getElementById("vote-submit-btn")]);
+
+            let current_row = 0;
+            let current_col = 0;
+
+            function clamp(min, val, max) {
+                if (val < min) return min;
+                else if (val > max) return max;
+                else return val;
+            }
+
+            function select_current(clickOnLabel=true) {
+                things_to_decide[current_row][current_col].focus();
+                if (clickOnLabel && things_to_decide[current_row][current_col].tagName === "LABEL")
+                    things_to_decide[current_row][current_col].click();
+            }
+
+            select_current();
+
+            const KEYCODES = {
+                "tab": 9,
+                "left": 37,
+                "right": 39,
+                "up": 38,
+                "down": 40,
+            };
+
+            document.getElementById("student-vote-form").addEventListener("keydown", e => {
+                if (things_to_decide[current_row][current_col].tagName == "TEXTAREA"
+                    && e.key != "Tab") {
+                    return;
+                }
+
+                console.log(e);
+                switch(e.key) {
+                    case "Tab":
+                        if (e.shiftKey) current_row--;
+                        else current_row++;
+                        current_row = clamp(0, current_row, things_to_decide.length - 1);
+
+                        // Focus the selected answer on the newly selected row
+                        let selected = things_to_decide[current_row].findIndex(el => el.classList.contains("active"));
+                        if (selected === -1) {
+                            selected = Math.floor((things_to_decide[current_row].length - 1) / 2);
+                        }
+                        current_col = selected;
+                        select_current();
+                        e.preventDefault();
+                        return;
+                    case "ArrowLeft":
+                        current_col--;
+                        break;
+                    case "ArrowRight":
+                        current_col++;
+                        break;
+                    case "ArrowUp":
+                    case "ArrowDown":
+                        e.preventDefault();
+                        return;
```

Comment:
So you can not use Up and Down to go to the next / previous line?

Reply:
I am not sure about this yet. Right now, if you tab to the next row, you either land where your current selection is or you start on the neutral option. I feel like the up and down arrow keys should not trigger any horizontal movement, so they would just move up or down in the grid. There are two options for what happens next:
1) The element you land on is selected: This sucks, because you can't go through the page with the arrow keys without overwriting your selections and you would have "made a choice" (selected an option) just by moving on to the next question. Some people might be used to scroll with the arrow keys, so it would be awful for them too.
2) The element you land on is focused, but not selected: This is weird as well, since the other operation with the arrow keys automatically select the element you move on. Now, you have to either press space to select the focussed element, or move away and back on to the focussed element with the left / right key.
I think both options are unintuitive, which is why I disabled the arrow keys for now.

Additionally, in text areas, you _can_ use the left and right key, since these are not needed in this situation for anything else (and are pretty needed for text manipulation). I would want to have the arrow keys in text areas as well, but they would be reserved for navigation.

Reply:
I'd vote in favor of option 1: You need to select one entry in each row anyway, so selecting the same choice as in the previous answer (that is, "go straight down and select that entry") seems reasonable for me. If some element is already selected, obviously, we shouldn't overwrite that choice, but rather jump to that position. Then, I think this is a good implementation that makes sense for someone going through the questions.

(Also, I think having some consistent behavior on the arrow keys that might not be intuitive to all users is still better than completely disabling the arrow keys at all)

IMHO, inside text areas, all arrow keys should just move around in the text field. TAB should be the only key allowing you to leave it (but I think your current code already has this early check for text fields?)

Reply:
You are right, having at least the option to use up / down without limiting the text area controls shouldn't hurt anyone except people scrolling with the arrow keys. Now that I think about it, this shouldn't be a problem either, considering that the site scrolls automatically if done as proposed in the other comments :)
Yeah, I think all key events that are not the tab key go into the text area already, but I am not standing on the hacked together code as of now. Right now I want to find out what works good for using the site and I can figure out how to do that later
Thanks for your feedback! :smiley:

Reply:
I would prefer if we could get around to imitate tab-behaviour. The default behaviour, like on `master` currently has these issues:

1. <kbd>Tab</kbd> works as expected for changing the current focus to the next item. However, in Firefox, tabbing forwards on a radio group without selected item will focus the next item of the same group. I hope we can solve this without mocking the tabbing to much, maybe only capture <kbd>Tab</kbd> when on a radio item with no item selected in its group.
2. We can’t select a radio button by it’s number. Well, this is something we definitively have to do custom, 
but it _should_ be possible without capturing <kbd>Tab</kbd>. (Nit: I would also put “No Answer” to <kbd>0</kbd>)
3. Tabbing will focus the name of a contributor on a collapsible, which I can’t really see the use-case for it. Maybe adding a `tabindex` or similar.

So, I would only concentrate on the number hot-keys (with a nice keboard legend like for the quick review, so it is discoverable) and as secondary goal to unify the behaviour of <kbd>Tab</kbd> for unselected radio groups.
I hope this will save you from some edge-cases when handling the tabbing and could keep your code small and clean :)

Reply:
I was working on an implementation that gives control back to the browser if the focused item is in a row where something is selected, but I noticed two downsides with this approach:
1. The browser always focuses the first element ("strongly agree"). I think it would be best if, when tabbing onto a row where no option is selected, the "neutral" option would be focused first.
2. I would still have to figure out where the next element that would be focused is, since I want to center it with `scrollIntoView`. I could also hack this with a timeout or maybe another event (keyup or so), but it would be easier to do right away.

The focusing of the contributor collapse was easy to remove with `tabindex` though, thanks for that hint!